src/modules/exchange/echange.abstract.ts

```diff
@@ -0,0 +1,87 @@
+import axios from 'axios';
+import { HttpException, HttpStatus } from '@nestjs/common';
+
+import { RateResponce } from './dto/exchange.dto';
+
+export enum Provider {
+  ExchangeRate = 'ExchangeRate',
+  PrivatBank = 'PrivatBank',
+  MonoBank = 'MonoBank',
+}
+
+export interface ExchangeRateData {
+  conversion_rates: {
+    USD: number;
+  };
+}
+
+export interface PrivatBankData {
+  ccy: string;
+  base_ccy: string;
+  buy: string;
+  sale: string;
+}
+
+export interface MonoBankData {
+  currencyCodeA: number;
+  currencyCodeB: number;
+  rateBuy: number;
+  rateSell: number;
+}
+
+type CurrencyData = ExchangeRateData | PrivatBankData[] | MonoBankData[];
+
+export abstract class BaseExchangeService {
```

Comment:
You used abstract but with implemented methods in it. It should be non-abstract class then. As maximum you can use some abstract properties or methods.

Reply:
If I make it non abstract- then I can't add abstract getUsdUahRate function inside, which I would like to do.
Can you pls suggest a solution for this?